### PR TITLE
feat(jsx): normalize class attribute

### PR DESCRIPTION
**Pull request by kasperskei**
Class binding like [vue.js](https://vuejs.org/guide/essentials/class-and-style.html#binding-html-classes).

### CODE REVIEW

File: packages/jsx/src/index.test.tsx [hunk 1]
```diff
@@ -303,7 +303,7 @@ it('custom component', setup((ctx, h, hf, mount, parent) => {
   const Component = (props: JSX.HTMLAttributes) => <div {...props} />
 
   assert.instance(<Component />, window.HTMLElement)
-  assert.is(((<Component draggable />) as HTMLElement).draggable, true)
+  assert.is(((<Component draggable="true" />) as HTMLElement).draggable, true)
```
Comment (kasperskei) [inline packages/jsx/src/index.test.tsx:306]: Attribute is not boolean. That's why the tests failed.
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/draggable

File: packages/jsx/src/index.ts [hunk 2]
```diff
@@ -156,46 +189,47 @@ export const reatomJsx = (ctx: Ctx, DOM: DomApis = globalThis.window) => {
...
+      } else if (k === 'ref') {
+        ctx.schedule(() => {
+          const cleanup = prop(ctx, element)
+          if (typeof cleanup === 'function') {
+            let list = unsubscribesMap.get(element)
+            if (!list) unsubscribesMap.set(element, (list = []))
+            unlink(element, () => cleanup(ctx, element))
+          }
+        })
+      } else if (k === 'class' && containsAtom(prop)) {
```
Comment (artalar) [inline packages/jsx/src/index.ts]:
```suggestion
      } else if ((isAtom(prop) && !prop.__reatom.isAction) || ((k === 'class' || k === 'className') && isObject(prop) && (isAtom(prop) || (prop = reatomClassName(prop))))) {
```

Also

```ts
const reatomClassName = (value: object) =>
  atom((ctx, state = '') => {
    if (state && ctx.cause.pubs.length === 0) {
      ctx.cause.proto.computer = null
    } else {
      state = ''
      for (const name in value) {
        if (isAtom(value[name]) ? ctx.spy(value[name]) : value[name]) state += name + ' '
      }
    }

    return state
  })
```

File: packages/jsx/src/utils.test.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    })), '')
+  })
+
+  it('handles falsy array correctly', () => {
+    assert.is(ctx.get(cn([
+      '',
+      null,
+      undefined,
+      {},
+      [],
```
Comment (Akiyamka) [inline packages/jsx/src/utils.test.ts:38]: It's a bit inconsistent, that we treat `{}` and `[]`   
In js world they truly, and in objects of this lib - too.

File: packages/jsx/src/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      c: true,
+      d: {},
+      e: [],
+      f: atom(true),
+      g: () => undefined,
+    })), 'a b c d e f g ')
+  })
+
+  it('handles deep array correctly', () => {
+    assert.is(ctx.get(cn(['a', ['b', ['c']]])), 'a b c   ')
```
Comment (Akiyamka) [inline packages/jsx/src/utils.test.ts]: this extra spaces at the end - interned behavior?

File: packages/jsx/src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  if (typeof value === 'string') {
+    className = value
+  } else if (typeof value === 'function') {
+    className = reatomClassName(ctx, value(ctx))
+  } else if (Array.isArray(value)) {
+    const length = value.length
+    for (let i = 0; i < length; i++) {
+      const normalized = reatomClassName(ctx, value[i])
+      if (normalized) {
+        className += normalized + ' '
```
Comment (Akiyamka) [inline packages/jsx/src/utils.ts]:
```suggestion
        className += normalized !== '' ? normalized + ' ' : normalized
```

Comment (Akiyamka) [reply packages/jsx/src/utils.ts]: Should fix extra spaces

File: packages/jsx/src/utils.test.ts (hunk 4 above)
Comment (artalar) [reply packages/jsx/src/utils.test.ts]: Yep, it our optimization to not trying to trim each string

File: packages/jsx/src/index.ts [hunk 6]
```diff
@@ -110,12 +112,14 @@ export const reatomJsx = (ctx: Ctx, DOM: DomApis = globalThis.window) => {
         styleId = styles[val] = random().toString()
         stylesheet.innerText += '[data-reatom="' + styleId + '"]{' + val + '}\n'
       }
-      /** @see https://www.measurethat.net/Benchmarks/Show/11819/0/dataset-vs-setattribute */
+      /** @see https://www.measurethat.net/Benchmarks/Show/11819 */
       element.setAttribute('data-reatom', styleId)
-    } else if (key === 'style' && typeof val === 'object') {
```
Comment (artalar) [inline packages/jsx/src/index.ts:115]: Wow, why you removed the style props handling? Please, restore it and add additional tests for css vars, which will ensure that we will not remove this feature again later :)

File: packages/jsx/src/utils.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  | undefined
+  | Array<ClassValue>
+  | Atom<ClassValue>
+  | Record<string, AtomMaybe<string | number | boolean | null | undefined | object>>
+  | ((ctx: CtxSpy) => ClassValue)
+>
+
+export const cn = (value: ClassValue): Atom<string> => atom((ctx) => reatomClassName(ctx, value))
+
+const reatomClassName = (ctx: CtxSpy, value: ClassValue): string => {
```
Comment (artalar) [inline packages/jsx/src/utils.ts]: A thing name starts with "reatom*" if it returns an atom or an action. Name it `parseClasses`, maube, or something like that

File: packages/jsx/src/utils.test.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+  it('handles complex correctly', () => {
+    assert.is(ctx.get(cn(() => atom(() => [
+      'a',
+      {b: atom(true)},
+      ['c'],
+      atom('d'),
+      () => 'e',
+    ]))), 'a b  c  d e ')
+  })
+})
```
Comment (artalar) [inline packages/jsx/src/utils.test.ts:86]: Please, test the change behaviour of created atom to ensure that the created atom correctly subscribes (spy) to nested atom. You can change one of the existing tests

File: packages/jsx/src/utils.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { atom, Atom, AtomMaybe, CtxSpy, isAtom } from '@reatom/core'
+import { isObject} from '@reatom/utils'
+
+type ClassValue = AtomMaybe<
```
Comment (artalar) [inline packages/jsx/src/utils.ts]:
```suggestion
export type ClassNameValue = AtomMaybe<
```

File: packages/jsx/src/utils.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+const reatomClassName = (ctx: CtxSpy, value: ClassValue): string => {
+  let className = ''
+  while (isAtom(value)) value = ctx.spy(value)
+  if (typeof value === 'string') {
+    className = value
+  } else if (typeof value === 'function') {
+    className = reatomClassName(ctx, value(ctx))
+  } else if (Array.isArray(value)) {
+    const length = value.length
+    for (let i = 0; i < length; i++) {
```
Comment (artalar) [inline packages/jsx/src/utils.ts]:
```suggestion
    for (let i = 0; i < value.length; i++) {
```

File: packages/jsx/src/utils.ts [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+      const normalized = reatomClassName(ctx, value[i])
+      if (normalized) {
```
Comment (artalar) [inline packages/jsx/src/utils.ts]: Always use strict comparison in this kind utilities to improve performance
```suggestion
      if (normalized !== '') {
```

**Review by artalar — CHANGES_REQUESTED** [review]

File: packages/jsx/src/index.ts [hunk 12]
```diff
@@ -125,8 +127,15 @@ export const reatomJsx = (ctx: Ctx, DOM: DomApis = globalThis.window) => {
...
       if (val == null || val === false) element.removeAttribute(key)
-      else if (val === true) element.setAttribute(key, '')
-      else element.setAttribute(key, String(val))
+      else {
+        val = val === true ? '' : String(val)
+        /**
+         * @see https://measurethat.net/Benchmarks/Show/54
+         * @see https://measurethat.net/Benchmarks/Show/31249
+         */
+        if (key === 'class' && element instanceof HTMLElement) element.className = val
```
Comment (artalar) [inline packages/jsx/src/index.ts:137]: Should we handle className too? Btw, don't forget to change the types
```suggestion
        if ((key === 'class' || key === 'className') && element instanceof HTMLElement) element.className = val
```

**Review by artalar — APPROVED** [review]

**Pull request merged**